src/Taboo/TabooRoom.py

```diff
@@ -40,15 +50,110 @@ def postProcessConnect(self, ws):
+        if self.playerByWs.get(ws, None):
+            try:
+                self.playerByWs[ws].delConn(ws)
+            except KeyError:
+                pass
+
+        try:
+            del self.playerByWs[ws]
+        except KeyError:
+            pass
+
         self.publishGiStatus()
 
     def processMsg(self, qmsg):
-        """Handle messages from the queue coming to this room"""
+        """Handle messages from the queue coming to this room
+
+            message-type1: ["JOIN", ...]
+        """
         if super(TabooRoom, self).processMsg(qmsg):
             return True
 
+        if qmsg.jmsg[0] == "JOIN":
+            return self.__processJoin(qmsg)
+
         return False
+
+    def __processJoin(self, qmsg):
+        """
+        ["JOIN", playerName, team:int={0..T}]
+        """
+        ws = qmsg.initiatorWs
+
+        if ws not in self.playerByWs or self.playerByWs[ws]: #QQ
+            self.txQueue.put_nowait(ClientTxMsg(["JOIN-BAD",
+                                                 "Unexpected JOIN message from client that "
+                                                 "has already joined"], {ws}, initiatorWs=ws))
+            return True
+
+
+        if len(qmsg.jmsg) != 3:
+            self.txQueue.put_nowait(ClientTxMsg(["JOIN-BAD", "Invalid message length"],
+                                                {ws}, initiatorWs=ws))
+            return True
+
+        _, playerName, teamNumber = qmsg.jmsg
+
+        if not isinstance(playerName, str) or not validPlayerNameRe.match(playerName):
+            self.txQueue.put_nowait(ClientTxMsg(["JOIN-BAD", "Invalid player name", playerName],
+                                                {ws}, initiatorWs=ws))
+            return True
+
+        if (not isinstance(teamNumber, int) or
+            teamNumber < 0 or teamNumber >= self.hostParameters.numTeams):
+            self.txQueue.put_nowait(ClientTxMsg(["JOIN-BAD", "Invalid team number", teamNumber],
+                                                {ws}, initiatorWs=ws))
+            return True
+
+        player, team = self.getPlayer(playerName)
+
+        # A new player joins
+        if not player:
+            if teamNumber not in self.teams:
+                #A new team
+                self.teams[teamNumber] = TabooTeam(teamNumber)
+            team = self.teams.get(teamNumber)
+            player = TabooPlayer(self.txQueue, playerName, team)
+            self.__finalize_join(ws, player, team.teamNumber)
+            #self.processEvent(PlayerJoin(player))
+            return True
+
+        assert team is not None, "What happened!"
+        #new join-request for an existing player will be accepted
+        #but we ignore the requested team and associate the player to the original team
+        if teamNumber != team.teamNumber:
+            trace(Level.warn, "Player {} forced to join their "
+            "original team {}".format(player.name, team.teamNumber))
```

Comment:
how about move all this in a function such as:

   return self.joinPlayer(playerName, teamNumber)

and let it do all this logic (including finalize?)?

Reply:
done